File: packages/crypto/src/session/serialization.ts
import { bytesToBase64, base64ToBytes } from '../encoding';
import type { RatchetState } from '../ratchet/double-ratchet';

/**
 * A JSON-safe mirror of `RatchetState` (Uint8Arrays → base64, Map → array of
 * entries) — this is what actually gets wrapped (storage/wrap.ts) and persisted to
 * IndexedDB, and what's reconstituted back into a live `RatchetState` on load. Kept
 * as its own module so the "how do we serialize a ratchet" question has exactly one
 * answer, reused by every caller instead of ad hoc `JSON.stringify` at call sites.
 */
export interface SerializedRatchetState {
  dhSelfPrivateKey: string;
  dhSelfPublicKey: string;
  dhRemote: string | null;
  rootKey: string;
  chainKeySend: string | null;
  chainKeyRecv: string | null;
  sendCount: number;
  recvCount: number;
  previousSendCount: number;
  skippedMessageKeys: Array<[string, string]>;
}

export function serializeRatchetState(state: RatchetState): SerializedRatchetState {
  return {
    dhSelfPrivateKey: bytesToBase64(state.dhSelf.privateKey),
    dhSelfPublicKey: bytesToBase64(state.dhSelf.publicKey),
    dhRemote: state.dhRemote ? bytesToBase64(state.dhRemote) : null,
    rootKey: bytesToBase64(state.rootKey),
    chainKeySend: state.chainKeySend ? bytesToBase64(state.chainKeySend) : null,
    chainKeyRecv: state.chainKeyRecv ? bytesToBase64(state.chainKeyRecv) : null,
    sendCount: state.sendCount,
    recvCount: state.recvCount,
    previousSendCount: state.previousSendCount,
    skippedMessageKeys: Array.from(state.skippedMessageKeys.entries()).map(([k, v]) => [k, bytesToBase64(v)]),
  };
}

export function deserializeRatchetState(serialized: SerializedRatchetState): RatchetState {
  return {
    dhSelf: {
      privateKey: base64ToBytes(serialized.dhSelfPrivateKey),
      publicKey: base64ToBytes(serialized.dhSelfPublicKey),
    },
    dhRemote: serialized.dhRemote ? base64ToBytes(serialized.dhRemote) : null,
    rootKey: base64ToBytes(serialized.rootKey),
    chainKeySend: serialized.chainKeySend ? base64ToBytes(serialized.chainKeySend) : null,
    chainKeyRecv: serialized.chainKeyRecv ? base64ToBytes(serialized.chainKeyRecv) : null,
    sendCount: serialized.sendCount,
    recvCount: serialized.recvCount,
    previousSendCount: serialized.previousSendCount,
    skippedMessageKeys: new Map(serialized.skippedMessageKeys.map(([k, v]) => [k, base64ToBytes(v)])),
  };
}

File: apps/mobile/lib/crypto/session/serialization.dart
/// JSON-safe mirror of `RatchetState` — direct port of
/// `packages/crypto/src/session/serialization.ts`. This is what actually gets
/// wrapped (crypto/storage/wrap.dart) and persisted, and what's reconstituted back
/// into a live `RatchetState` on load.
library;

import '../encoding.dart';
import '../primitives.dart';
import '../ratchet/double_ratchet.dart';

class SerializedRatchetState {
  final String dhSelfPrivateKey;
  final String dhSelfPublicKey;
  final String? dhRemote;
  final String rootKey;
  final String? chainKeySend;
  final String? chainKeyRecv;
  final int sendCount;
  final int recvCount;
  final int previousSendCount;
  final List<List<String>> skippedMessageKeys; // [ [id, base64], ... ]

  const SerializedRatchetState({
    required this.dhSelfPrivateKey,
    required this.dhSelfPublicKey,
    required this.dhRemote,
    required this.rootKey,
    required this.chainKeySend,
    required this.chainKeyRecv,
    required this.sendCount,
    required this.recvCount,
    required this.previousSendCount,
    required this.skippedMessageKeys,
  });

  Map<String, dynamic> toJson() => {
    'dhSelfPrivateKey': dhSelfPrivateKey,
    'dhSelfPublicKey': dhSelfPublicKey,
    'dhRemote': dhRemote,
    'rootKey': rootKey,
    'chainKeySend': chainKeySend,
    'chainKeyRecv': chainKeyRecv,
    'sendCount': sendCount,
    'recvCount': recvCount,
    'previousSendCount': previousSendCount,
    'skippedMessageKeys': skippedMessageKeys,
  };

  static SerializedRatchetState fromJson(Map<String, dynamic> json) {
    return SerializedRatchetState(
      dhSelfPrivateKey: json['dhSelfPrivateKey'] as String,
      dhSelfPublicKey: json['dhSelfPublicKey'] as String,
      dhRemote: json['dhRemote'] as String?,
      rootKey: json['rootKey'] as String,
      chainKeySend: json['chainKeySend'] as String?,
      chainKeyRecv: json['chainKeyRecv'] as String?,
      sendCount: json['sendCount'] as int,
      recvCount: json['recvCount'] as int,
      previousSendCount: json['previousSendCount'] as int,
      skippedMessageKeys: (json['skippedMessageKeys'] as List)
          .map((e) => (e as List).map((v) => v as String).toList())
          .toList(),
    );
  }
}

SerializedRatchetState serializeRatchetState(RatchetState state) {
  return SerializedRatchetState(
    dhSelfPrivateKey: bytesToBase64(state.dhSelf.privateKey),
    dhSelfPublicKey: bytesToBase64(state.dhSelf.publicKey),
    dhRemote: state.dhRemote != null ? bytesToBase64(state.dhRemote!) : null,
    rootKey: bytesToBase64(state.rootKey),
    chainKeySend: state.chainKeySend != null ? bytesToBase64(state.chainKeySend!) : null,
    chainKeyRecv: state.chainKeyRecv != null ? bytesToBase64(state.chainKeyRecv!) : null,
    sendCount: state.sendCount,
    recvCount: state.recvCount,
    previousSendCount: state.previousSendCount,
    skippedMessageKeys: state.skippedMessageKeys.entries.map((e) => [e.key, bytesToBase64(e.value)]).toList(),
  );
}

RatchetState deserializeRatchetState(SerializedRatchetState s) {
  return RatchetState(
    dhSelf: KeyPairBytes(privateKey: base64ToBytes(s.dhSelfPrivateKey), publicKey: base64ToBytes(s.dhSelfPublicKey)),
    dhRemote: s.dhRemote != null ? base64ToBytes(s.dhRemote!) : null,
    rootKey: base64ToBytes(s.rootKey),
    chainKeySend: s.chainKeySend != null ? base64ToBytes(s.chainKeySend!) : null,
    chainKeyRecv: s.chainKeyRecv != null ? base64ToBytes(s.chainKeyRecv!) : null,
    sendCount: s.sendCount,
    recvCount: s.recvCount,
    previousSendCount: s.previousSendCount,
    skippedMessageKeys: {for (final pair in s.skippedMessageKeys) pair[0]: base64ToBytes(pair[1])},
  );
}
